expression/builtin_compare.go

```diff
@@ -1070,68 +1070,91 @@ func isTemporalColumn(expr Expression) bool {
 }
 
 // tryToConvertConstantInt tries to convert a constant with other type to a int constant.
-func tryToConvertConstantInt(ctx sessionctx.Context, isUnsigned bool, con *Constant) (_ *Constant, isAlwaysFalse bool) {
+func tryToConvertConstantInt(ctx sessionctx.Context, targetFieldType *types.FieldType, con *Constant) (
+	correctVal *Constant, // The computed result. If the constant can convert to int, return the val. Else return 'con'(the input).
+	exceptionalVal *Constant,
+	// The val is nil or else indicates whether the int column [cmp] const is always true/false.
+	// It is used to get more information about why the 'con' is always true/false.
+	// If the op == LT,LE,GT,GE and it gets an Overflow when converting, return inf/-inf.
+	// If the op == EQ,NullEQ and the constant can never be equal to the int column, return the constant.
+	//	(eg. "integer  =  1.1" is definitely false.)
+	// Otherwise, return nil.
+) {
 	if con.GetType().EvalType() == types.ETInt {
-		return con, false
+		return con, nil
 	}
 	dt, err := con.Eval(chunk.Row{})
 	if err != nil {
-		return con, false
+		return con, nil
 	}
 	sc := ctx.GetSessionVars().StmtCtx
-	fieldType := types.NewFieldType(mysql.TypeLonglong)
-	if isUnsigned {
-		fieldType.Flag |= mysql.UnsignedFlag
-	}
-	dt, err = dt.ConvertTo(sc, fieldType)
+
+	dt, err = dt.ConvertTo(sc, targetFieldType)
 	if err != nil {
-		return con, terror.ErrorEqual(err, types.ErrOverflow)
+		if terror.ErrorEqual(err, types.ErrOverflow) {
+			return con, &Constant{
+				Value:   dt,
+				RetType: targetFieldType,
+			}
+		}
+		return con, nil
 	}
 	return &Constant{
 		Value:        dt,
-		RetType:      fieldType,
+		RetType:      targetFieldType,
 		DeferredExpr: con.DeferredExpr,
-	}, false
-}
-
-// RefineComparedConstant changes an non-integer constant argument to its ceiling or floor result by the given op.
-// isAlwaysFalse indicates whether the int column "con" is false.
-func RefineComparedConstant(ctx sessionctx.Context, isUnsigned bool, con *Constant, op opcode.Op) (_ *Constant, isAlwaysFalse bool) {
+	}, nil
+}
+
+// RefineComparedConstant changes a non-integer constant argument to its ceiling or floor result by the given op.
+func RefineComparedConstant(ctx sessionctx.Context, targetFieldType *types.FieldType, con *Constant, op opcode.Op) (
```

Comment:
We'd better define `targetFieldType` as `types.FieldType` instead `*types.FieldType` to avoid the column.FieldType be modified unexpectedly